relay-server/src/actors/spooler/mod.rs

```diff
@@ -501,18 +394,330 @@ impl BufferService {
+            .await
+            .and_then(|r| r.try_get(0))
+            .map_err(BufferError::FileSizeReadFailed)?;
+
+        relay_statsd::metric!(histogram(RelayHistograms::BufferDiskSize) = size as u64);
+        Ok(size)
+    }
+
+    /// Returns `true` if the maximum size is reached, `false` otherwise.
+    async fn is_full(&self) -> Result<bool, BufferError> {
+        let current_size = self.estimate_spool_size().await?;
+        Ok(current_size as usize >= self.max_disk_size)
+    }
+
+    /// Returns `true` if the spool is empty, `false` otherwise.
+    async fn is_empty(&self) -> Result<bool, BufferError> {
+        let is_empty = sql::select_one()
+            .fetch_optional(&self.db)
+            .await
+            .map_err(BufferError::FetchFailed)?
+            .is_none();
+
+        Ok(is_empty)
+    }
+
+    /// Enqueues data into on-disk spool.
+    async fn enqueue(
+        &self,
+        key: QueueKey,
+        managed_envelope: ManagedEnvelope,
+    ) -> Result<(), BufferError> {
+        let received_at = managed_envelope.received_at().timestamp_millis();
+        sql::insert(
+            key,
+            managed_envelope.into_envelope().to_vec().unwrap(),
+            received_at,
+        )
+        .execute(&self.db)
+        .await
+        .map_err(BufferError::InsertFailed)?;
+
+        relay_statsd::metric!(counter(RelayCounters::BufferWrites) += 1);
+        Ok(())
+    }
+}
+
+/// The state which defines the [`BufferService`] behaviour.
+#[derive(Debug)]
+enum BufferState {
+    /// Only memory buffer is enabled.
+    Memory { ram: InMemory },
+
+    /// The memory buffer is used, but also disk spool is configured.
+    ///
+    /// The disk will be used when the memory limit will be hit.
+    MemoryFileStandby { ram: InMemory, disk: OnDisk },
+
+    /// THe disk is full and we use the memory to save incoming envelopes as much as we can. But
+    /// the disk still used to read from.
+    MemoryFileRead { ram: InMemory, disk: OnDisk },
+
+    /// Only disk used for read/write operations.
+    File { disk: OnDisk },
+}
+
+impl BufferState {
+    /// Creates the initial [`BufferState`] depending on the provided config.
+    ///
+    /// If the on-disk spool is configured, the `BufferState::MemoryDiskStandby` will be returned or
+    /// `BufferState::Memory` otherwise.
+    fn init(max_memory_size: usize, on_disk: Option<OnDisk>) -> Self {
+        let in_memory = InMemory {
+            buffer: BTreeMap::new(),
+            max_memory_size,
+            used_memory: 0,
+        };
+        match on_disk {
+            Some(on_disk) => Self::MemoryFileStandby {
+                ram: in_memory,
+                disk: on_disk,
+            },
+            None => Self::Memory { ram: in_memory },
+        }
+    }
+
+    /// Creates `BufferState::MemoryDiskStandby` from provided disk reference and maximum in-memory
+    /// buffer size.
+    fn to_memory(max_memory_size: usize, disk: &OnDisk) -> Self {
```

Comment:
Regardless of what names we choose for the states, can we align the names of the `to_` methods with the state names? In this case, that would be `to_memory_file_standby`.